### PR TITLE
[versioning] [EOS-24955] GetObject with version

### PR DESCRIPTION
**Pull request by sjain09**
# Problem Statement

- Get Latest object data if versioning is either enabled or suspended and no version ID is specified in the request.
- Get object data for a specific version if version ID is specified in the request for a version enabled or suspended bucket.
- If wrong versionId is specified, return error - InvalidArgument.

# Design
https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/766902714/GetObject+LLD

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [x] Unit and System Tests are added
- [x] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: server/s3_error_codes.cc
Comment (indrajitzagade) [inline server/s3_error_codes.cc:60]: Please post a link of amazon s3 documentation regarding including the invalid arguments in the error response.

File: st/clitests/awss3api_spec.py
Comment (indrajitzagade) [inline st/clitests/awss3api_spec.py:1938]: *with empty

File: server/s3_error_codes.cc
Comment (sjain09) [reply server/s3_error_codes.cc:60]: I could not find any document that describes error response in detail. This is purely based on observation.

File: st/clitests/awss3api_spec.py
Comment (sjain09) [reply st/clitests/awss3api_spec.py:1938]: "empty" is a place holder for empty version id. Internally it is getting translated into empty version id inside get_object function. I have made it in this way to differentiate between None and "" (empty string) values for version id that is being passed into get_object() fuction.

Comment (sjain09) [reply st/clitests/awss3api_spec.py:1938]: I will update the description too.

Comment (sjain09) [reply st/clitests/awss3api_spec.py]: done

File: server/s3_get_object_action.cc
Comment (ajay-paratmandali) [inline server/s3_get_object_action.cc:843]: low priority change

can we use 
set_version_id()

"x-amz-version-id" this field related to object metadata 
and it will be good if we can hide it from action

File: server/s3_error_codes.cc
Comment (indrajitzagade) [reply server/s3_error_codes.cc:60]: ok, please post the observation for info.

**Review by ajay-paratmandali — APPROVED** [review]

**Comment by indrajitzagade** [comment]
@t7ko-seagate Please review this PR. We are making some changes to the base classes, I think your review would help.

**Comment by indrajitzagade** [comment]
@sjain09 Please run Jenkins jobs and share the results.

File: server/s3_error_codes.cc
Comment (keithpine) [reply server/s3_error_codes.cc:60]: We should be documenting all of these kinds of observations somewhere. Not sure if in LLD, or a general document of errors.

File: server/action_base.h
Comment (keithpine) [inline server/action_base.h:159]: Why is a map necessary? Do you have an example where there are multiple arguments returned?

File: server/s3_error_codes.cc
Comment (keithpine) [inline server/s3_error_codes.cc:39]: std::move on a constant has no effect. Either get rid of the std::move calls, or change the args to values.

File: server/s3_get_object_action.cc
Comment (keithpine) [inline server/s3_get_object_action.cc:104]:
```suggestion
             "Empty versionId provided in the request\n");
```

File: server/action_base.h
Comment (sjain09) [reply server/action_base.h:159]: I don't have such examples or may be there are none as we should return error as soon as we detect invalid argument. Now I feel it's not necessary yet it act as a container to hold the invalid argument. Should I go ahead and replace it with two different strings to hold arg name and arg value?

File: server/s3_get_object_action.cc
Comment (sjain09) [reply server/s3_get_object_action.cc:843]: I think it should be fine to do it here as we are setting all other metadata attribute values here.

File: server/s3_error_codes.cc
Comment (sjain09) [reply server/s3_error_codes.cc]: removed std::move().

File: server/s3_get_object_action.cc
Comment (sjain09) [reply server/s3_get_object_action.cc]: done

**Comment by sjain09** [comment]
Removed  ST related to suspended bucket case as this is not supported in this PI.

**Comment by sjain09** [comment]
> @sjain09 Please run Jenkins jobs and share the results.

@indrajitzagade , Jenkins Jobs will require PutObject API changes to be merged first. These changes are required in STs.

**Comment by keithpine** [comment]
I'm getting a code format error when I try to build this?

```text
[ERROR:Code formatting]
One or more modified files do not comply with clang-format.
Running below command to see required changes:
        'git clang-format --style=Google --extensions=c,cc,h,java --diff --commit HEAD~1'

====================================================================




diff --git a/ut/s3_get_object_action_test.cc b/ut/s3_get_object_action_test.cc
index 105cd334..2e1203df 100644
--- a/ut/s3_get_object_action_test.cc
+++ b/ut/s3_get_object_action_test.cc
@@ -262,7 +262,7 @@ TEST_F(S3GetObjectActionTest, FetchObjectInfoFailedEmptyVersionId) {
   CREATE_BUCKET_METADATA;
 
   EXPECT_CALL(*(bucket_meta_factory->mock_bucket_metadata), get_state())
-     .WillRepeatedly(Return(S3BucketMetadataState::present));
+      .WillRepeatedly(Return(S3BucketMetadataState::present));
   EXPECT_CALL(*ptr_mock_request, set_out_header_value(_, _)).Times(AtLeast(1));
   EXPECT_CALL(*ptr_mock_request, send_response(_, _)).Times(1);
   EXPECT_CALL(*(bucket_meta_factory->mock_bucket_metadata),
@@ -293,7 +293,7 @@ TEST_F(S3GetObjectActionTest, FetchObjectInfoFailedInvalidVersionId) {
   CREATE_OBJECT_METADATA;
 
====================================================================
Code formatting check...Failed
```

**Comment by indrajitzagade** [comment]
> > @sjain09 Please run Jenkins jobs and share the results.
> 
> @indrajitzagade , Jenkins Jobs will require PutObject API changes to be merged first. These changes are required in STs.

Sure

**Comment by sjain09** [comment]
> I'm getting a code format error when I try to build this?
> 
> ```
> [ERROR:Code formatting]
> One or more modified files do not comply with clang-format.
> Running below command to see required changes:
>         'git clang-format --style=Google --extensions=c,cc,h,java --diff --commit HEAD~1'
> 
> ====================================================================
> 
> 
> 
> 
> diff --git a/ut/s3_get_object_action_test.cc b/ut/s3_get_object_action_test.cc
> index 105cd334..2e1203df 100644
> --- a/ut/s3_get_object_action_test.cc
> +++ b/ut/s3_get_object_action_test.cc
> @@ -262,7 +262,7 @@ TEST_F(S3GetObjectActionTest, FetchObjectInfoFailedEmptyVersionId) {
>    CREATE_BUCKET_METADATA;
>  
>    EXPECT_CALL(*(bucket_meta_factory->mock_bucket_metadata), get_state())
> -     .WillRepeatedly(Return(S3BucketMetadataState::present));
> +      .WillRepeatedly(Return(S3BucketMetadataState::present));
>    EXPECT_CALL(*ptr_mock_request, set_out_header_value(_, _)).Times(AtLeast(1));
>    EXPECT_CALL(*ptr_mock_request, send_response(_, _)).Times(1);
>    EXPECT_CALL(*(bucket_meta_factory->mock_bucket_metadata),
> @@ -293,7 +293,7 @@ TEST_F(S3GetObjectActionTest, FetchObjectInfoFailedInvalidVersionId) {
>    CREATE_OBJECT_METADATA;
>  
> ====================================================================
> Code formatting check...Failed
> ```

Fixed code formatting issue.

File: server/s3_error_codes.cc
Comment (sjain09) [reply server/s3_error_codes.cc]: Updated the observation on page - https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/817889369/Observations

File: server/action_base.h
Comment (sjain09) [reply server/action_base.h]: Replaced std::map with two strings to hold arg name and arg value.

Comment (keithpine) [reply server/action_base.h]: You could also use an std::pair, but it's fine.

File: server/s3_get_object_action.cc
Comment (keithpine) [inline server/s3_get_object_action.cc:101]:
```suggestion
               request->get_query_string_value("versionId").empty()) {
```

File: server/s3_object_action_base.cc
Comment (keithpine) [inline server/s3_object_action_base.cc:97]:
```suggestion
             (request->get_query_string_value("versionId").empty())) {
```

File: server/s3_object_metadata.cc
Comment (keithpine) [inline server/s3_object_metadata.cc:424]:
```suggestion
  if (object_version_id.empty()) {
```

File: server/s3_object_metadata.cc
Comment (keithpine) [inline server/s3_object_metadata.cc:443]:
```suggestion
      (!requested_object_version_id.empty() &&
```

File: server/s3_object_metadata.cc
Comment (keithpine) [inline server/s3_object_metadata.cc:453]:
```suggestion
    } else if (requested_object_version_id.empty()) {
```

**Review by keithpine — CHANGES_REQUESTED** [review]
Sorry, I thought I left these comments previously. Please prefer `std::string::empty()` over comparison to `""`.

File: st/clitests/awss3api_spec.py
Comment (indrajitzagade) [inline st/clitests/awss3api_spec.py:1928]: But we still specify the version_id here?

Comment (sjain09) [reply st/clitests/awss3api_spec.py]: Updated, thanks for pointing out.

File: server/s3_get_object_action.cc
Comment (sjain09) [reply server/s3_get_object_action.cc]: done.

File: server/s3_object_action_base.cc
Comment (sjain09) [reply server/s3_object_action_base.cc]: done.

File: server/s3_object_metadata.cc
Comment (sjain09) [reply server/s3_object_metadata.cc]: done.

File: server/s3_object_metadata.cc
Comment (sjain09) [reply server/s3_object_metadata.cc]: done.

**Review by keithpine — COMMENTED** [review]
LGTM, but we can't merge this until the Put Object change is merged, is that right? The system tests would fail, no? Can we disable the ST and merge?

**Comment by keithpine** [comment]
Jenkins Results:

1. [Pre-Merge Test](http://eos-jenkins.colo.seagate.com/job/S3server/job/S3Server-PreMerge-Test/736/) - PASS
2. [Deployment S3 Server](http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/CentOS-7.9.2009/job/s3server/377/) - PASS
    1. [Downstream PR-Deployment](http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Cortx-Deployment/job/CentOS-3.9.2009/job/pr-deployment/1281/) - UNSTABLE
4. [Mini-Provisioning](http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/Mini-Provisioning/job/CentOS-7.9.2009/job/S3-Server/679/) - PASS

**Pull request merged**
